// convex/dist/esm/cli/lib/deployApi/modules.js
"use strict";
import { z } from "zod";
export const moduleEnvironment = z.union([
  z.literal("isolate"),
  z.literal("node")
]);
export const moduleConfig = z.object({
  path: z.string(),
  source: z.string(),
  sourceMap: z.optional(z.string()),
  environment: moduleEnvironment
});
export const nodeDependency = z.object({
  name: z.string(),
  version: z.string()
});
export const udfConfig = z.object({
  serverVersion: z.string(),
  // RNG seed encoded as Convex bytes in JSON.
  importPhaseRngSeed: z.any(),
  // Timestamp encoded as a Convex Int64 in JSON.
  importPhaseUnixTimestamp: z.any()
});
export const sourcePackage = z.any();
export const visibility = z.union([
  z.object({ kind: z.literal("public") }),
  z.object({ kind: z.literal("internal") })
]);
export const analyzedFunction = z.object({
  name: z.string(),
  pos: z.any(),
  udfType: z.union([
    z.literal("Query"),
    z.literal("Mutation"),
    z.literal("Action")
  ]),
  visibility: z.nullable(visibility),
  args: z.nullable(z.string()),
  returns: z.nullable(z.string())
});
export const analyzedModule = z.object({
  functions: z.array(analyzedFunction),
  httpRoutes: z.any(),
  cronSpecs: z.any(),
  sourceMapped: z.any()
});
