// rakkasjs/dist/build-ID62FRV7.js
import {
  doPrerender
} from "./chunk-WYYMNBQG.js";
import {
  cleanOptions
} from "./chunk-3VG2XVTH.js";
import {
  version
} from "./chunk-TTMWBCHJ.js";

// src/cli/build.ts
import { multibuild } from "@vavite/multibuild";
import pico from "picocolors";
async function build(root, options) {
  const buildOptions = cleanOptions(options);
  let config;
  let total;
  let viteSteps;
  let paths;
  let adapter;
  function logStep(index, name) {
    config.logger.info(
      "\n" + pico.magenta("rakkas") + ": " + name + " (" + pico.green(`${index}/${total}`) + ")"
    );
  }
  process.env.NODE_ENV = options.mode || "production";
  await multibuild(
    {
      root,
      base: options.base,
      mode: options.mode,
      configFile: options.config,
      logLevel: options.logLevel,
      clearScreen: options.clearScreen,
      build: buildOptions
    },
    {
      onInitialConfigResolved(resolvedConfig) {
        config = resolvedConfig;
        config.logger.info(
          pico.black(pico.bgMagenta(" RAKKAS ")) + " " + pico.magenta(version) + " \u{1F483}"
        );
        total = viteSteps = config.buildSteps?.length || 1;
        paths = config.api?.rakkas?.prerender || [];
        if (paths.length) {
          total += 1;
        }
        adapter = config.api?.rakkas?.adapter;
        if (adapter.bundle) {
          total += 1;
        }
      },
      onStartBuildStep(info) {
        logStep(info.currentStepIndex + 1, "Building " + info.currentStep.name);
      }
    }
  );
  let step = viteSteps + 1;
  if (paths.length) {
    logStep(step++, "Prerendering static routes");
    await doPrerender(config, void 0, true);
  }
  if (adapter.bundle) {
    logStep(step++, `Adapting for ${adapter.name}`);
    await adapter.bundle(config.root);
  }
}
export {
  build
};
