// rakkasjs/dist/index.server.js
import {
  ClientOnly,
  ClientSuspense,
  DefaultErrorPage,
  ErrorBoundary,
  Head,
  Link,
  Redirect,
  ResponseHeaders,
  StyledLink,
  cancelLastNavigation,
  navigate,
  prefetchRoute,
  queryOptions,
  requestContextStorage,
  runServerSideMutation,
  runServerSideQuery,
  useErrorBoundary,
  useFormMutation,
  useHead,
  useLocation,
  useMutation,
  useMutations,
  useNavigationBlocker,
  usePageContext,
  useQuery,
  useQueryClient,
  useRequestContext,
  useRouteParams,
  useServerSentEvents,
  useServerSideMutation,
  useServerSideQuery,
  useSubmit
} from "./chunk-53EWDOCE.js";
import "./chunk-JSBRDJBE.js";

// src/features/async-local-request-context/lib-server.ts
function getRequestContext() {
  const ctx = requestContextStorage?.getStore();
  if (!ctx) {
    throw new Error("Request context not found");
  }
  return ctx;
}
export {
  ClientOnly,
  ClientSuspense,
  DefaultErrorPage,
  ErrorBoundary,
  Head,
  Link,
  Redirect,
  ResponseHeaders,
  StyledLink,
  cancelLastNavigation,
  getRequestContext,
  navigate,
  prefetchRoute,
  queryOptions,
  runServerSideMutation as runSSM,
  runServerSideQuery as runSSQ,
  runServerSideMutation,
  runServerSideQuery,
  useErrorBoundary,
  useFormMutation,
  useHead,
  useLocation,
  useMutation,
  useMutations,
  useNavigationBlocker,
  usePageContext,
  useQuery,
  useQueryClient,
  useRequestContext,
  useRouteParams,
  useServerSentEvents as useSSE,
  useServerSideMutation as useSSM,
  useServerSideQuery as useSSQ,
  useServerSentEvents,
  useServerSideMutation,
  useServerSideQuery,
  useSubmit
};
